### ISSUE TITLE
Explanation Gender % and Entropy

### ISSUE THREAD

**Issue by CodeCoder20** [tae898/age-gender#14]
Hi Kim, 
thanks for your nice algorithm.
I am using it for gender classification and I'm a little bit confused with the interpretation of the output variables.

1. If a Face is flagged with 100% Female, does that mean, that the facial embeddings are to 100% similar to a gender typical facial embedding? Or how do you interpret this value?
2. Entropy (Shannon Entropy?) is a measure of uncertainty.
So if I would like to set a cutoff value, that the algorithm has to reach to accept the prediciton, e.g. 95% should I look at the entropy or at your percentage value?

Hope you can help, looking forward to your answer. Thanks in advance.

**Comment by tae898** [comment]
Hey, 

Thank you for your interest in my work. Below is my answer to your questions.

During training, what I do is a typical-supervised learning for a classification problem. That is, we want to train a function approximator $f(y|x, \theta)$. This is a neural network whose input is $x$ (e.g., image) and whose output is $y$. $y$ is typically a categorical distribution. In our case, we only have two classes (i.e., male or female), so $y$ reduces to Bernoulli distribution.

We optimize for the cross entropy loss. That is, when the neural network outputs $\hat{y} = f(x)$, we want $\hat{y}$ to be as close as possible to its supervised label $y$. Note that the labels $y$ are ALWAYS one-hot encoded vectors. In our cases, they are always either $[1.0, 0]^T$, $[0, 1.0]^T$. We know that the neural network tries to generalize. Let's say that at test inference time, given a test data point $x_{test}^{(i)}$, it outputs $\hat{y}^{(i)}$. We are “tempted” to interpret the value of this vector as probability. For example, if it gives us $[female=1.0, male=0.0]^T$, we want to interpret it as "the neural network is 100% sure that this face is a female face. This must be a VERY female face, then.". Or, if it gives us $[female=0.5, male=0.5]^T$, we might want to say "this face is somewhere between a very female and a very male face.". It's hard to say such a thing. The neural network has never seen a label such as $[female=0.5, male=0.5]^T$, but how can it magically learn to generalize to such hypothetical data points?

This is why we use Bayesian inference: $p(y|x) =\int_{\theta} p(y|x, \theta) p(\theta)d\theta$. Instead of finding a point estimate of $\theta^{MLE}$ using maximum likelihood estimation, we take advantage of the distribution of $\theta$: $p(\theta)$, and get $p(y|x)$ by marginalizing out $\theta$. $p(y|x)$ is more "accurate" than $p(y|x, \theta^{MLE})$, since it accounts for the uncertainty of $\theta$. However, this is not easy to compute. Computing $p(y|x) =\int_{\theta} p(y|x, \theta) p(\theta)d\theta$ is often intractable or very expensive. 

And here's why I used MC dropout. Although I used Dropout in a traditional way to get $\theta^{MLE}$ during training, I followed the MC dropout paper that still turns on Dropout at test phase so that the neural network becomes stochastic, instead of deterministic. Let's say that I ran $N$ forward passes through the neural network with the Dropout turned on. This means that we are sampling $\theta \sim p(\theta)$ $N$ times and computing $p(y|x, \theta)$ $N$ times. Now if we compute the expectation (i.e., mean) of the $N$ values of $p(y|x, \theta)$, it'll probably be very close to $p(y|x, \theta^{MLE})$. But now the best part is that we have $N$ values of $p(y|x, \theta)$. We can compute the "uncertainty" of $p(y|x, \theta)$ by using these values. We can compute either variance or entropy, which can measure uncertainty.

Unfortunately, you have to set your own threshold value for this uncertainty metric. There is no clear rule for this. One thing you can do is you make your own held-out dataset, compute the uncertainty of each sample, and decide yourself which cut-off value would be the best for your own setup.

Hope it helps.

**Comment by CodeCoder20** [comment]
Thank you very much, that helps a lot.
